### ISSUE TITLE
Terminal go to recent directory should standardize path formats

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#157932]
Currently various paths formats are added depending on the shell. For example pwsh, wsl(bash) and git bash are shown below for the same folder:

![image](https://user-images.githubusercontent.com/2193314/184205858-2b40e24d-e02e-4755-ac07-eab74b29bae8.png)

We should convert the paths into a URI and store them in a consistent format, then when go to recent directory is triggered the path can be changed back to the right format for the shell in question.

This is technically an all platform problem but in practice it would probably only be seen on Windows.

**Comment by babakks** [comment]
@Tyriar I've submitted a PR to address this issue. However, it's just a minimalistic implementation that merely unifies the visual path formatting.

On the other hand, the ideal scenario I can think of is as follows: By executing the `Go To Recent Directory...`, I see the item `/C:/Github/microsoft/vscode-docs` (or something like that) on the picker without any duplicates (in other formats). When I select it, no matter what shell type I use at the moment (i.e., `GitBash`, `WSL` or `CMD`/`PWSH`), a `cd` command to that directory appears in the shell and successfully changes the CWD to it. I mean, on `GitBash`. goes to `/c/...` or on `WSL` goes to `/mnt/c/...`.

Is this what you actually had in mind @Tyriar?

**Comment by Tyriar** [comment]
@babakks yep that's what I was thinking, in other words take the shell type into account when sending it. We do have some wsl helpers in the preparePathForShell helper (I think that'd the right name) which I just hooked up so part of that might be working now actually 🤔

**Comment by babakks** [comment]
I don't work on this anymore, so please feel free to contribute to this issue.

@Tyriar I've submitted another PR for this issue, but, as I've mentioned in the description, it's not a ready-to-review one. That's just to help anyone who wants to work on this. So feel free to close it if you should. 

I close the first PR since that's not a viable solution.

**Comment by Tyriar** [comment]
Cross posting from https://github.com/microsoft/vscode/pull/158267

---

This is getting there but it's a little different from what I think we should be doing which is:

- When the directories are added to the list, store them in as a URI (`URI.toString()`?) to ensure they're all consistent, this includes setting a remote authority when necessary.
- When it comes time to use the directories, convert them to the shell type's format. This is mostly happening in the PR but we should make it a little more generic. For example creating a function like this:
  ```ts
  interface ITerminalBackend {
    formatUriForShellType(path: URI, shellType: TerminalShellType): Promise<string> | string;
  }
  ```
  This function would only go to the backend if it's actually needed, for example I know wsl path is needed due to some edge cases, but can cygpath be evaluated purely based on the drive letter?
